add documentation for multi-project metrics (#213)
GoogleCloudPlatform/k8s-stackdriver: custom-metrics-stackdriver-adapter/examples/cross-project-metrics/README.md

# Custom Metrics with Stackdriver

This is a minimal example showing how to use the stackdriver adapter to get metrics from
a different project than the one your cluster is running in. By default, the adapter
uses the credentials of the node machine. There are three approaches for granting permissions
to your other project:

1. Grant the default service account for your project access to `roles/monitoring.viewer` in
   your other project.

> 1. Create your GKE cluster with a service account, and grant the service account access to 

Reviewer: ```suggestion
1. Create your GKE cluster with a custom service account, and grant the service account access to 
```

   `roles/monitoring.viewer` in your other project

1. (Recommended) Create a service a service account for the stackdriver-adapter, and mount

>    the credentials into the pod. This is the method this example implements.

Reviewer: ```suggestion
   the credentials into the pod. This is the method explained in this example.
```

This example assumes you already have custom metrics set up in your cluster

## Create a service account with credentials

For this example, you will need to create a service account. See [Creating a service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts)

> for how to do that.

Reviewer: ```suggestion
for instructions.
```

> Next, you will need to create a key for that services account. See [Creating service account keys](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#iam-service-account-keys-create-console) for how to do that.

Reviewer: ```suggestion
Next, you will need to create a key for that services account. See [Creating service account keys](https://cloud.google.com/iam/docs/creating-managing-service-account-keys#iam-service-account-keys-create-console) for instructions.
```

> Copy the key to this directory. The filename should be `key.json`

Reviewer: ```suggestion
Copy the key to your current directory. The filename should be `key.json`
```

> Add `roles/monitoring.viewer` to your service account in all projects that you want it to grab metrics from.

Reviewer: Maybe also add a link? https://cloud.google.com/iam/docs/granting-changing-revoking-access

## Creating the kubernetes secret

Run the following command:

```
kubectl create secret generic custom-metrics-stackdriver-adapter --from-file=./key.json -n custom-metrics
```

## Apply the modified adapter

Run the following command:

```

> kubectl apply -f modified-adapter.yml

Reviewer: Can you add a link using raw github content instead of assuming that user has cloned this repo? Something like: https://raw.githubusercontent.com/GoogleCloudPlatform/k8s-stackdriver/master/custom-metrics-stackdriver-adapter/examples/cross-project-metrics/adapter.yaml

```

## Picking a metric to scale on

Open `cross-project-hpa.yml`. You'll notice that the metrics are based on the number of undelivered messages on a
pubsub subscription called `my-subscription` in a different project called `my-other-project`. Modify these selectors
to match your environment and save the file. 

## Apply the deployment and hpa

Run the following commands:

```
kubectl apply -f deployment-to-be-scaled.yml
kubectl apply -f cross-project-hpa.yml
```

## Check that it's working

Run the following command:

```
kubectl describe hpa dummy
```

You should see that the hpa is able to detect the metric in your other project, and is scaling your deployment accordingly.

## Troubleshooting

There are several places where cross project autoscaling can get stuck, so when troubleshooting
make sure to check them one by one.

> 1. Check that your service account has the right permissions. It needs `roles/monitoring.viewer`

Reviewer: ```suggestion
1. Check that your service account has the right permissions on the IAM page of your other project. It needs `roles/monitoring.viewer`
```

2. Check that custom-metrics-stackdriver-adapter is able to read the metric from Stackdriver

Run kubeproxy 
```
kubectl proxy -p 8001
```
And check the available metrics
```

> curl http://localhost:8001/apis/custom.metrics.k8s.io/v1beta1/

Reviewer: Unfortunately this won't work. This returns metrics from your current project, while the requested metric can be specified only for the other project. It may still make sense to check that for predefined metrics (non-custom metrics) that should exist for all projects, such as pubsub metrics used here.
